File: sorting/bubble/bubble.js
/*
In Place Bubble Sort

Sort array in place by iterating through the array and swapping adjacent elements if they are out of order.
Do this repeatedly until the array is sorted.

Set a flag to indicate whether the array is sorted.

Iterate through the array
check if current element is greater than the next element
  if so, swap them and set flag to false
  if not, continue
at the end of the iteration, if the flag is set to false, then continue iterating
at the end of the iteration, if the flag is set to true, then the array is sorted
*/
const array = [2, 1, 4, 7, 3, 6, 5];

const bubbleSort = (array) => {
  let sorted = false;
  while (!sorted) {
    sorted = true;
    for (let i = 1; i < array.length; i++) {
      if (array[i - 1] > array[i]) {
        [array[i], array[i - 1]] = [array[i - 1], array[i]]
        sorted = false;
      }
    }
  }
  return array
};

console.log(bubbleSort(array))

File: sorting/bubble/bubble.py
"""
In Place Bubble Sort

Set flag to indicate whether the array is sorted or not -- initialize to true;

Iterate through the array:
    Check current element and adjacent element
      If adjacent element is greater than current element, swap them and set flag to false
      If adjacent element is less than or equal to current element, do nothing
After iterating through the array, check the flag -- if true, the array is sorted and can be returned
if false, repeate the process with a while loop checking the flag
"""


mylist = [5, 3, 1, 2, 4]

def bubble_sort(mylist):
  """
  In place bubble sort
  """
  sorted = False
  while not sorted:
    sorted = True
    for i in range(1, len(mylist)):
      if mylist[i - 1] > mylist[i]:
        sorted = False
        mylist[i - 1], mylist[i] = mylist[i], mylist[i - 1]
  return mylist

print(bubble_sort(mylist))
